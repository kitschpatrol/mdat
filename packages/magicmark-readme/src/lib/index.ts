export {
	type CheckReadmeFileOptions,
	type CheckReadmeFileReport,
	type CheckReadmeStringOptions,
	type CheckReadmeStringReport,
	checkReadmeFile,
	checkReadmeString,
} from './check'

export {
	type ExpandReadmeFileOptions,
	type ExpandReadmeFileReport,
	type ExpandReadmeStringOptions,
	type ExpandReadmeStringReport,
	expandReadmeFile,
	expandReadmeString,
} from './expand'
