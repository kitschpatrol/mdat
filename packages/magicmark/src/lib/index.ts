export {
	type CheckFileReport,
	type CheckOptions,
	type CheckReport,
	checkAst,
	checkFile,
	checkFiles,
	checkString,
} from './check'
export {
	type ExpandAstOptions,
	type ExpandAstReport,
	type ExpandFileOptions,
	type ExpandFileReport,
	type ExpandFilesOptions,
	type ExpandStringOptions,
	type ExpandStringReport,
	expandAst,
	expandFile,
	expandFiles,
	expandString,
} from './expand'
export { default as log } from './log'
export { type Rules, getSoleRule } from './rules'
export { logCheckReport, logExpandFilesReport } from './utilities'
