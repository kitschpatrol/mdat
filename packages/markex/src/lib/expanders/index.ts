import readme from './readme'
import { type ExpanderPreset } from './types'

export default {
	readme,
} satisfies Record<string, ExpanderPreset>
