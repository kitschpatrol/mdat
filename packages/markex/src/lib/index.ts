export { type ExpandAstOptions, type ExpandStringOptions, expandAst, expandString } from './expand'
export { default as presets } from './rules'
export type { Rule, RuleSet } from './rules/types'
export {
	type ValidateAstOptions,
	type ValidateStringOptions,
	validateAst,
	validateString,
} from './validate'
