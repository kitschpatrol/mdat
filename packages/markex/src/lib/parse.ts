import json5 from 'json5'
import type { JsonObject } from 'type-fest'

export function parseCommentText(
	text: string,
): { args: JsonObject | undefined; keyword: string } | undefined {
	// Get the keyword with args
	const match = /^\s*<!--\/*[\s#-]*(.+)\s*-->\s*$/.exec(text)?.at(1)?.trim()
	if (match === undefined) return undefined

	// Get the args
	const parts = /^([^\s({]+)[\s()]*({.+})?\)?$/g.exec(match)
	if (parts === null) return undefined

	const keyword = parts.at(1)
	if (keyword === undefined) return undefined

	const rawArgs = parts.at(2)

	// Use json5 so we can be more forgiving about unquoted keys
	const args = rawArgs ? json5.parse<JsonObject>(rawArgs) : undefined

	return { args, keyword }
}
