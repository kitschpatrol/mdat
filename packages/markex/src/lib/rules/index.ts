import readme from './readme'
import { type RuleSet } from './types'

export default {
	readme,
} satisfies Record<string, RuleSet>
