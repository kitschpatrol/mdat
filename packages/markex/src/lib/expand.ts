import { parseCommentText } from './parse'
import { type Rule, type RuleSet } from './rules/types'
import { type Html, type Root } from 'mdast'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'
import type { JsonObject } from 'type-fest'
import { CONTINUE, EXIT, visit } from 'unist-util-visit'

// String
export type ExpandStringOptions = ExpandAstOptions
export async function expandString(
	markdown: string,
	options: ExpandStringOptions,
): Promise<{ expandedString: string; report: string[] }> {
	const ast = remark().use(remarkGfm).parse(markdown)
	const { expandedAst, report } = await expandAst(ast, options)

	return {
		expandedString: remark().use(remarkGfm).stringify(expandedAst),
		report,
	}
}

// AST
export type ExpandAstOptions = {
	expansionRules: RuleSet
	keywordPrefix?: string
	meta?: boolean
}
export async function expandAst(
	ast: Root,
	options: ExpandAstOptions,
): Promise<{ expandedAst: Root; report: string[] }> {
	const { expansionRules, keywordPrefix = '', meta = false } = options

	// Extract template expansion commands from comment nodes
	// https://github.com/syntax-tree/mdast/blob/main/readme.md

	// Happens in two passes to accommodate async rules

	// Save promises as we go
	const newContent: Array<{
		applySequence: number
		args: JsonObject | undefined
		getContent: Rule['getContent']
		openingComment: Html
	}> = []

	// TODO stay at top level
	// First pass adds closing comments if needed and removes existing children
	visit(ast, 'html', (node, index, parent) => {
		if (parent !== undefined && index !== undefined) {
			// Parse the comment contents
			const result = parseCommentText(node.value)
			if (result === undefined) return CONTINUE
			const { args, keyword } = result

			// Look for a matching rule in the rule set
			const matchingRule = Object.values(expansionRules).find(
				(rule) => `${keywordPrefix}${rule.keyword}` === keyword,
			)
			if (matchingRule === undefined) return CONTINUE

			// The keyword already contains the prefix from the parser
			// Look for a closing closing comment, if there is one we'll delete everything
			// between the opening and closing comments
			const closingTagIndex = getClosingTagIndex(ast, index, keyword)
			const tagsToReplace = closingTagIndex ? closingTagIndex - index - 1 : 0
			parent.children.splice(index + 1, tagsToReplace)

			// Add closing tag if it doesn't exist
			if (closingTagIndex === undefined) {
				const closingNode: Html = {
					type: 'html',
					value: `<!-- /${keyword} -->`,
				}
				parent.children.splice(index + 1, 0, closingNode)
			}

			// Save the reference to promise function and its args
			// to generate new nodes later on
			newContent.push({
				applySequence: matchingRule.applicationOrder ?? 0,
				args,
				getContent: matchingRule.getContent,
				openingComment: node,
			})
		}
	})

	// Sort newContent in place to apply expansion rules in the order they're received in the array
	newContent.sort((a, b) => a.applySequence - b.applySequence)

	// Execution, not just promise resolution, must be deferred to here
	// to ensure table of contents has all generated headings
	for (const { args, getContent, openingComment } of newContent) {
		const newMarkdownString = await getContent(ast, args)
		const newNodes = remark().use(remarkGfm).parse(newMarkdownString).children
		const openingCommentIndex = ast.children.indexOf(openingComment)
		ast.children.splice(openingCommentIndex + 1, 0, ...newNodes)
	}

	// Add or remove the meta comment if requested
	const message = 'Warning: Content in HTML comment blocks generated by markex'
	// Remove existing meta comments
	ast.children = ast.children.filter(
		(node) => !(node.type === 'html' && node.value.startsWith(`<!-- ${message}`)),
	)

	if (meta) {
		const date = new Date().toISOString().slice(0, 10)
		const metaComment: Html = {
			type: 'html',
			value: `<!-- ${message} on ${date}. -->`,
		}
		ast.children.unshift(metaComment)
	}

	return {
		expandedAst: ast,
		report: newContent.map(({ openingComment }) => openingComment.value),
	}
}

// Helpers
function getClosingTagIndex(
	ast: Root,
	startFromIndex: number,
	keywordWithPrefix: string,
): number | undefined {
	let matchingIndex: number | undefined
	visit(ast, 'html', (node, index, parent) => {
		if (parent !== undefined && index !== undefined && index >= startFromIndex) {
			const result = parseCommentText(node.value)
			if (result === undefined) return CONTINUE

			if (`/${keywordWithPrefix}` === result.keyword) {
				matchingIndex = index
				return EXIT
			}
		}
	})
	return matchingIndex
}
