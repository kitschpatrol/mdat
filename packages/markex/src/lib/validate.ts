/* eslint-disable max-depth */
/* eslint-disable complexity */
import { type ExpandAstOptions, type ExpandStringOptions } from './expand'
import { parseCommentText } from './parse'
import { type Rule } from './rules/types'
import chalk from 'chalk'
import { type Root } from 'mdast'
import plur from 'plur'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'

export type ValidateStringOptions = ExpandStringOptions
export async function validateString(
	markdown: string,
	options: ValidateStringOptions,
): Promise<Error[] | true> {
	const ast = remark().use(remarkGfm).parse(markdown)
	return validateAst(ast, options)
}

export type ValidateAstOptions = ExpandAstOptions
export async function validateAst(ast: Root, options: ValidateAstOptions): Promise<Error[] | true> {
	const { expansionRules, keywordPrefix } = options
	const errors: Error[] = []

	// Get valid expander comments, verify valid args
	const validExpanders: Rule[] = []
	for (const node of ast.children) {
		if (node.type === 'html') {
			const parsedComment = parseCommentText(node.value)
			if (parsedComment === undefined) continue
			const { args, keyword: commentKeyword } = parsedComment

			const matchingExpander = Object.values(expansionRules).find(
				(rule) => `${keywordPrefix ?? ''}${rule.keyword}` === commentKeyword,
			)

			if (matchingExpander === undefined) continue

			// Valid command, check args
			if (args) {
				try {
					await matchingExpander.getContent(ast, args)
				} catch (error) {
					if (error instanceof Error) {
						errors.push(error)
					}
				}
			}

			validExpanders.push(matchingExpander)
		}
	}

	// Check for missing required rules
	const absenceErrors: Error[] = []
	for (const rule of Object.values(expansionRules)) {
		if (rule.required && !validExpanders.includes(rule)) {
			absenceErrors.push(new Error(`  "${chalk.yellow(`<-- ${rule.keyword} -->`)}"`))
		}
	}

	if (absenceErrors.length > 0) {
		errors.push(
			new Error(
				chalk.bold.red(
					`${absenceErrors.length} required ${plur('comment', absenceErrors.length)} ${absenceErrors.length > 1 ? 'are' : 'is'} missing from the document:`,
				),
			),
			...absenceErrors,
		)
	}

	// Check for order issues
	const validExpandersWithOrder = validExpanders.filter((rule) => rule.order !== undefined)
	if (validExpandersWithOrder.length > 1) {
		// Force unwrap because we've checked for it in the filter above
		const sortedValidExpanders = [...validExpandersWithOrder].sort((a, b) => a.order! - b.order!)

		const sortErrors: Error[] = []

		for (const sortedExpander of validExpandersWithOrder) {
			const correctIndex = sortedValidExpanders.indexOf(sortedExpander)
			const currentIndex = validExpandersWithOrder.indexOf(sortedExpander)
			if (correctIndex !== currentIndex) {
				const up = correctIndex < currentIndex
				let message = `\t${up ? '↑' : '↓'} "${chalk.yellow(sortedExpander.keyword)}" should move ${up ? 'up' : 'down'} so it comes`

				if (correctIndex === 0) {
					message = `${message} before "${chalk.yellow(sortedValidExpanders[correctIndex + 1].keyword)}"`
				} else if (correctIndex === validExpandersWithOrder.length - 1) {
					message = `${message} after "${chalk.yellow(sortedValidExpanders[correctIndex - 1].keyword)}"`
				} else {
					message = `${message} after "${chalk.yellow(sortedValidExpanders[correctIndex - 1].keyword)}" and before "${chalk.yellow(sortedValidExpanders[correctIndex + 1].keyword)}"`
				}

				sortErrors.push(new Error(message))
				continue
			}
		}

		if (sortErrors.length > 0) {
			errors.push(
				new Error(
					chalk.bold.red(
						`${sortErrors.length} ${plur('comment', sortErrors.length)} ${sortErrors.length > 1 ? 'are' : 'is'} not in the correct order:`,
					),
				),
				...sortErrors,
			)
		}

		// TODO maybe check for missing closing tags, indicating that the tool hasn't been run?
	}

	if (errors.length > 0) {
		return errors
	}

	return true
}
