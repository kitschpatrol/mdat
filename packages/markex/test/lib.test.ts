import { expandString, presets, validateString } from '../src/lib'
import { parseCommentText } from '../src/lib/parse'
import fs from 'node:fs/promises'
import { describe, expect, it } from 'vitest'

describe('expandString', () => {
	it('should expand comments and handle arguments', async () => {
		const markdown = await fs.readFile('./test/assets/readme-basic.md', 'utf8')
		const { expandedString } = await expandString(markdown, { expansionRules: presets.readme })
		expect(expandedString).toMatchSnapshot()
	})

	it('should expand special header and footer comments', async () => {
		const markdown = await fs.readFile('./test/assets/readme-header-footer.md', 'utf8')
		const { expandedString } = await expandString(markdown, { expansionRules: presets.readme })
		expect(expandedString).toMatchSnapshot()
	})

	it('should expand prefixed comments', async () => {
		const markdown = await fs.readFile('./test/assets/readme-basic-prefixed.md', 'utf8')
		const { expandedString } = await expandString(markdown, {
			expansionRules: presets.readme,
			keywordPrefix: 'tp.',
		})
		expect(expandedString).toMatchSnapshot()
	})
})

describe('basic parsing', () => {
	const basicResult = { args: undefined, keyword: 'title' }
	const basicResultPrefixed = { args: undefined, keyword: 'tp.title' }

	it('should not parse non-comments', () => {
		expect(parseCommentText('<!- title')).toEqual(undefined)
		expect(parseCommentText('<!!-- title() -->')).toEqual(undefined)
		expect(parseCommentText('title() -->')).toEqual(undefined)
	})

	it('should parse basic comments', () => {
		expect(parseCommentText('<!-- title -->')).toEqual(basicResult)
		expect(parseCommentText('<!-- title() -->')).toEqual(basicResult)
	})

	it('should forgive spacing variations', () => {
		expect(parseCommentText('<!--     title -->')).toEqual(basicResult)
		expect(parseCommentText('<!-- title-->')).toEqual(basicResult)
		expect(parseCommentText('<!--title -->')).toEqual(basicResult)
		expect(parseCommentText('<!--title-->')).toEqual(basicResult)
		expect(parseCommentText('<!--title()-->')).toEqual(basicResult)
		expect(parseCommentText('<!--title (  )-->')).toEqual(basicResult)
		expect(parseCommentText('<!-- title (  )  -->')).toEqual(basicResult)
		expect(parseCommentText('<!--     title-->')).toEqual(basicResult)
		expect(parseCommentText('<!--title-->')).toEqual(basicResult)
	})

	it('should forgive extra garbage in basic comments', () => {
		expect(parseCommentText('<!--// title -->')).toEqual(basicResult)
		expect(parseCommentText('<!--/// title -->')).toEqual(basicResult)
		expect(parseCommentText('<!--#title -->')).toEqual(basicResult)
		expect(parseCommentText('<!--####title -->')).toEqual(basicResult)
		expect(parseCommentText('<!-- #### title -->')).toEqual(basicResult)
	})

	it('should parse prefixed comments', () => {
		expect(parseCommentText('<!-- tp.title -->')).toEqual(basicResultPrefixed)
	})

	// TODO case handling
})

describe('option parsing', () => {
	const stringResult = { args: { prefix: 'ðŸ˜¬' }, keyword: 'title' }
	const numberResult = { args: { prefix: 1 }, keyword: 'title' }
	const booleanResult = { args: { prefix: true }, keyword: 'title' }

	it('should parse basic options', () => {
		expect(parseCommentText('<!-- title({prefix: "ðŸ˜¬"}) -->')).toEqual(stringResult)
		expect(parseCommentText('<!-- title({prefix: 1}) -->')).toEqual(numberResult)
		expect(parseCommentText('<!-- title({prefix: true}) -->')).toEqual(booleanResult)
	})

	it('should parse without parentheses', () => {
		expect(parseCommentText('<!-- title{prefix: "ðŸ˜¬"} -->')).toEqual(stringResult)
		expect(parseCommentText('<!-- title{prefix: 1} -->')).toEqual(numberResult)
		expect(parseCommentText('<!-- title{prefix: true} -->')).toEqual(booleanResult)
	})

	it('should forgive spacing variations', () => {
		expect(parseCommentText('<!-- title{prefix: "ðŸ˜¬"} -->')).toEqual(stringResult)
		expect(parseCommentText('<!--title{  prefix:   "ðŸ˜¬" }-->')).toEqual(stringResult)
		expect(parseCommentText('<!-- title {prefix: 1}-->')).toEqual(numberResult)
		expect(parseCommentText('<!--title   {prefix: true} -->')).toEqual(booleanResult)
		expect(parseCommentText('<!-- title({prefix: "ðŸ˜¬"}) -->')).toEqual(stringResult)
		expect(parseCommentText('<!--title({  prefix:   "ðŸ˜¬" })-->')).toEqual(stringResult)
		expect(parseCommentText('<!-- title ({prefix: 1})-->')).toEqual(numberResult)
		expect(parseCommentText('<!--title   ({prefix: true}) -->')).toEqual(booleanResult)
	})
})

describe('linting', () => {
	it('should not report errors when linted and valid', async () => {
		const markdown = await fs.readFile('./test/assets/readme-basic.md', 'utf8')
		const lintReport = await validateString(markdown, { expansionRules: presets.readme })

		expect(lintReport).toEqual(true)
	})

	it('should report errors when linted and invalid', async () => {
		const markdown = await fs.readFile('./test/assets/readme-basic-invalid.md', 'utf8')
		const lintReport = await validateString(markdown, { expansionRules: presets.readme })

		expect(lintReport).not.toBe(true)
		expect(lintReport).toHaveLength(7)
	})
})
