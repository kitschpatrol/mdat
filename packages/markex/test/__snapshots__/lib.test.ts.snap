// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expandString > should expand comments and handle arguments 1`] = `
"<!-- title {prefix: 'üìù '} -->

# üìù markex

<!-- /title -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package](https://img.shields.io/npm/v/markex.svg)](https://npmjs.com/package/markex)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**CLI tool and library for using comments as templates in markdown files.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

* [History!](#history)
  * [Discovery](#discovery)
  * [Name and symbol](#name-and-symbol)
  * [Planet X disproved](#planet-x-disproved)
* [Orbit](#orbit)
* [Contributing](#contributing)
* [License](#license)

<!-- /table-of-contents -->

A long description might say pluto is a dwarf planet in the Kuiper belt.

## History!

<!-- code {file: "./test/assets/expander-test.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
console.log('expander-test')

\`\`\`

<!-- /code -->

### Discovery

In the 1840s, Urbain Le Verrier used Newtonian mechanics to predict the position of‚Ä¶

### Name and symbol

The name Pluto is for the Roman god of the underworld, from a Greek epithet for Hades‚Ä¶

### Planet X disproved

Once Pluto was found, its faintness and lack of a viewable disc cast doubt‚Ä¶

## Orbit

Pluto's orbital period is about 248 years‚Ä¶

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/markex/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
"
`;

exports[`expandString > should expand prefixed comments 1`] = `
"<!-- tp.title -->

# markex

<!-- /tp.title -->

<!-- tp.badges -->

[![NPM Package](https://img.shields.io/npm/v/markex.svg)](https://npmjs.com/package/markex)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

<!-- /tp.badges -->

<!-- tp.short-description -->

**CLI tool and library for using comments as templates in markdown files.**

<!-- /tp.short-description -->

A long description might say pluto is a dwarf planet in the Kuiper belt.

## History

### Discovery

In the 1840s, Urbain Le Verrier used Newtonian mechanics to predict the position of‚Ä¶

### Name and symbol

The name Pluto is for the Roman god of the underworld, from a Greek epithet for Hades‚Ä¶

### Planet X disproved

Once Pluto was found, its faintness and lack of a viewable disc cast doubt‚Ä¶

## Orbit

Pluto's orbital period is about 248 years‚Ä¶

<!-- tp.contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/markex/issues) and pull requests are welcome.

<!-- /tp.contributing -->

<!-- tp.license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /tp.license -->
"
`;

exports[`expandString > should expand special header and footer comments 1`] = `
"<!-- header -->

# markex

[![NPM Package](https://img.shields.io/npm/v/markex.svg)](https://npmjs.com/package/markex)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

**CLI tool and library for using comments as templates in markdown files.**

<!-- /header -->

Pluto is a dwarf planet in the Kuiper belt.

## Contents

## History

### Discovery

In the 1840s, Urbain Le Verrier used Newtonian mechanics to predict the position of‚Ä¶

### Name and symbol

The name Pluto is for the Roman god of the underworld, from a Greek epithet for Hades‚Ä¶

### Planet X disproved

Once Pluto was found, its faintness and lack of a viewable disc cast doubt‚Ä¶

## Orbit

Pluto's orbital period is about 248 years‚Ä¶

<!-- footer -->

## Contributing

[Issues](https://github.com/kitschpatrol/markex/issues) and pull requests are welcome.

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /footer -->
"
`;
